core/templates/services/prevent-page-unload-event.service.ts

```diff
@@ -25,35 +25,50 @@ import { WindowRef } from 'services/contextual/window-ref.service';
 })
 export class PreventPageUnloadEventService {
   private listenerActive: boolean;
+  validation: Function;
+  _preventPageUnloadEventHandlerBind = undefined;
+
   constructor(private windowRef: WindowRef) {
     this.listenerActive = false;
+    this.validation = undefined;
   }
 
-  addListener(): void {
+  addListener(callback: Function = () => {
+    return true;
+  }): void {
     if (this.listenerActive) {
       return;
     }
+    this.validation = callback;
+    this._preventPageUnloadEventHandlerBind =
+      this._preventPageUnloadEventHandler.bind(null, this.validation);
     this.windowRef.nativeWindow.addEventListener(
-      'beforeunload', this._preventPageUnloadEventHandler);
+      'beforeunload', this._preventPageUnloadEventHandlerBind, true);
     this.listenerActive = true;
   }
 
   removeListener(): void {
+    if (!this.listenerActive) {
+      return;
+    }
     this.windowRef.nativeWindow.removeEventListener(
-      'beforeunload', this._preventPageUnloadEventHandler);
+      'beforeunload', this._preventPageUnloadEventHandlerBind, true);
     this.listenerActive = false;
   }
 
   private _preventPageUnloadEventHandler(
-      e: BeforeUnloadEvent): void {
-    // The preventDefault call is used to trigger a confirmation before leaving.
-    e.preventDefault();
-    // According to the specification, to show the confirmation dialog an
-    // event handler should call preventDefault() on the event. However note
-    // that not all browsers support this method. So returnValue is also used.
-    // The exact value in returnValue is not relevant, but it needs to be set
-    // in order to trigger a confirmation before leaving.
-    e.returnValue = '';
+      validation: Function, e: BeforeUnloadEvent): void {
+    if (validation()) {
```

Comment:
This function is introduced as a callback, right? If it's about validation then please do the validation before adding the event.

Reply:
Yes, this is a callback. After adding the callback it will keep checking if the ` e.returnValue = '';` should run or not. ` e.returnValue = '';` will trigger the prevent reload modal/alert